wire/netaddressv2.go

```diff
@@ -0,0 +1,576 @@
+func maxNetAddressV2Payload() uint32 {
+	// The timestamp takes up four bytes.
+	plen := uint32(4)
+
+	// The ServiceFlag is a varint and its maximum size is 9 bytes.
+	plen += 9
+
+	// The netID is a single byte.
+	plen += 1
+
+	// The largest address is 512 bytes. Even though it will not be a valid
+	// address, we should read and ignore it. The preceeding varint to
+	// store 512 bytes is 3 bytes long. This gives us a total of 515 bytes.
+	plen += 515
+
+	// The port is 2 bytes.
+	plen += 2
+
+	return plen
+}
+
+// isOnionCatTor returns whether a given ip address is actually an encoded tor
+// v2 address. The wire package is unable to use the addrmgr's IsOnionCatTor as
+// doing so would give an import cycle.
+func isOnionCatTor(ip net.IP) bool {
+	onionCatNet := net.IPNet{
+		IP:   net.ParseIP("fd87:d87e:eb43::"),
+		Mask: net.CIDRMask(48, 128),
+	}
+	return onionCatNet.Contains(ip)
+}
+
+// NetAddressV2 defines information about a peer on the network including the
+// last time it was seen, the services it supports, its address, and port. This
+// struct is used in the addrv2 message (MsgAddrV2) and can contain larger
+// addresses, like Tor. Additionally, it can contain any NetAddress address.
+type NetAddressV2 struct {
+	// Last time the address was seen. This is, unfortunately, encoded as a
+	// uint32 on the wire and therefore is limited to 2106. This field is
+	// not present in the bitcoin version message (MsgVersion) nor was it
+	// added until protocol version >= NetAddressTimeVersion.
+	Timestamp time.Time
+
+	// Services is a bitfield which identifies the services supported by
+	// the address. This is encoded in CompactSize.
+	Services ServiceFlag
+
+	// Addr is the network address of the peer. This is a variable-length
+	// address. Network() returns the BIP-155 networkID which is a uint8
+	// encoded as a string. String() returns the address as a string.
+	Addr net.Addr
+
+	// Port is the port of the address. This is 0 if the network doesn't
+	// use ports.
+	Port uint16
+}
+
+// HasService returns whether the specified service is supported by the
+// address.
+func (na *NetAddressV2) HasService(service ServiceFlag) bool {
+	return na.Services&service == service
+}
+
+// AddService adds a service to the Services bitfield.
+func (na *NetAddressV2) AddService(service ServiceFlag) {
+	na.Services |= service
+}
+
+// ToLegacy attempts to convert a NetAddressV2 to a legacy NetAddress. This
+// only works for ipv4, ipv6, or torv2 addresses as they can be encoded with
+// the OnionCat encoding. If this method is called on a torv3 address, nil will
+// be returned.
+func (na *NetAddressV2) ToLegacy() *NetAddress {
+	legacyNa := &NetAddress{
+		Timestamp: na.Timestamp,
+		Services:  na.Services,
+		Port:      na.Port,
+	}
+
+	switch a := na.Addr.(type) {
+	case *ipv4Addr:
+		legacyNa.IP = a.addr[:]
+	case *ipv6Addr:
+		legacyNa.IP = a.addr[:]
+	case *torv2Addr:
+		legacyNa.IP = a.onionCatEncoding()
+	case *torv3Addr:
+		return nil
```

Comment:
Return an error instead perhaps? Still getting through the commit, need to see how it's used downstream.